sklearn/preprocessing/_discretization.py

```diff
@@ -231,32 +261,71 @@ def fit(self, X, y=None, sample_weight=None):
 
         n_samples, n_features = X.shape
 
-        if sample_weight is not None and self.strategy == "uniform":
-            raise ValueError(
-                "`sample_weight` was provided but it cannot be "
-                "used with strategy='uniform'. Got strategy="
-                f"{self.strategy!r} instead."
-            )
-
         if self.subsample is not None and n_samples > self.subsample:
             # Take a subsample of `X`
+            # When resampling, it is important to subsample **with replacement** to
+            # preserve the distribution, in particular in the presence of a few data
+            # points with large weights.
             X = resample(
                 X,
-                replace=False,
+                replace=True,
                 n_samples=self.subsample,
                 random_state=self.random_state,
+                sample_weight=sample_weight,
             )
-
-        n_features = X.shape[1]
-        n_bins = self._validate_n_bins(n_features)
+            # Since we already used the weights when resampling when provided,
+            # we set them back to `None` to avoid accounting for the weights twice
+            # in subsequent operations to compute weight-aware bin edges with
+            # quantiles or k-means.
+            sample_weight = None
 
         if sample_weight is not None:
             sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
 
+        n_features = X.shape[1]
+        n_bins = self._validate_n_bins(n_features)
+
         bin_edges = np.zeros(n_features, dtype=object)
+
+        if self.strategy == "quantile" and self.quantile_method == "warn":
+            if sample_weight is None:
+                warnings.warn(
+                    "Defaulting to quantile method linear this will be changed to\
+                                  average_inverted_cdf in scikit-learn version 1.9",
+                    FutureWarning,
+                )
+                self.quantile_method = "linear"
```

Comment:
We should never change the public attributes set as constructor parameters in `fit`: instead, we should define a local variable `quantile_method` in `fit` and change that local variable only based on the other values:

```python
quantile_method = self.quantile_method
if self.strategy == "quantile" and quantile_method == "warn":
    if sample_weight is None:
        # issue warning here
        quantile_method = "linear"
    else:
        # issue other warning here
        quantile_method = "inverted_cdf"
...
```

The public attributes (without trailing underscores) are meant to only be changed by passing them as constructor parameters or via `set_params` (e.g. internally when doing a grid search). They should always represent what the user has configured and not the result of some internal operations that happen in fit.

If we really want to expose the result as a public attribute (for inspection purpose), we could name the result attribute as `self.quantile_method_` (in addition to the existing `self.quantile_method` that would be left unchanged). But since inspecting the value of this attribute is probably not that interesting, we can instead just use a local variable to avoid growing the public API surface needlessly.